_posts/2025-04-23-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,202 @@
+<2> Require authenticated tool access. See also how the main MCP SSE endpoint is secured in the <<configuration>> section below.
+<3> Use the injected `SecurityIdentity` to return a name.
+
+[[configuration]]
+=== Configuration
+
+Finally, let's configure our secure MCP server:
+
+[source,properties]
+----
+quarkus.http.auth.permission.authenticated.paths=/mcp/sse
+quarkus.http.auth.permission.authenticated.policy=authenticated
+----
+<1> Enforce an authenticated access to the main MCP SSE endpoint. See also how the tool is secured with an annotation in the <<tool>> section above, though you can also secure access to the tool by also listing an SSE tools endpoint in the configuration, for example: `quarkus.http.auth.permission.authenticated.paths=/mcp/sse,/mcp/messages/*`.
+
+What about the OIDC configuration ? It is provided for you in devmode by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak].  It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started immediately. You can also register a custom Keycloak realm.
+
+We are ready to test our secure MCP server, both in DevUI and with `curl`.
+
+== Start MCP SSE server
+
+Start the server in the dev mode:
+
+[source,shell]
+----
+mvn quarkus:dev
+----
+
+[[mcp-server-devui]]
+=== Access MCP SSE server in DevUI
+
+Go to http://localhost:8080/q/dev[Dev UI], find both MCP Server and OpenId Connect cards:
+
+image::mcp_server_oidc_devui.png[MCP Server and OIDC in DevUI,align="center"]
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+What about other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] ?
+No problems, you can login to them from OIDC DevUI as well, for example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0] - the only requirement is to update your provider application registration to allow callbacks to DevUI.
+
+Now, after logging in with `Keycloak`, copy the acquired access token using a provided copy button:
+
+image::login_and_copy_access_token.png[Login and copy access token,align="center"]
+
+Now go the MCP Server card, select the `Tools` option and choose the `user-name-provider` tool:
+
+image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]
+
+Paste the copied access token into the Tool's `Bearer token` field:
+
+image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]
+
+Make a tool call by selecting the `Call` action and get the response:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+=== Access MCP SSE server with curl
```

Comment:
Actually, here you explain it a bit better - but you need to introduce where is happens in the protocol.